### PR TITLE
add a new method and a function

### PR DESCRIPTION
**Pull request by Freakwill**
powermethod is a well known method to compute the eigenvalues.

define a function to_reciprocal_matrix(A). With the help of the function, we can use 0 in a preference matrix. it will be corrected to a positive number, in order to make A reciprocal. see yettelevision.json. It makes the writing of json convenient.

### CODE REVIEW

File: examples/yettelevision.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      [1, 9, 9, 1]
+    ],
+    "alternatives:rating": [
+      [1, 0, 0, 0],
+      [3, 1, 0, 0],
+      [7, 5, 1, 3],
+      [5, 3, 0, 1]
+    ]
+  }
+}
```
Comment (mishrabhinav) [inline examples/yettelevision.json]: Every file should have a newline at the end.

File: pyahp/utils.py [hunk 2]
```diff
@@ -21,3 +21,55 @@ def normalize_priorities(criteria_pr, global_pr):
...
+            elif A[i, j]==0:
+                if A[j, i] !=0:
+                    A[i, j] = 1/A[j, i]
+                else:
+                    for k in range(m):
+                        if A[i, k] !=0 and A[k, j] !=0:
+                            A[i, j] = A[i, k] * A[k, j]
+                            break
+                    else:
+                        raise ValueError('There are so many zeros! I am not able to convert A[%d,%d] to a vaild value!'%(i,j))
```
Comment (mishrabhinav) [inline pyahp/utils.py]: The error messages should not be in first person. Please create a new error in `errors.py` (similar to the ones in the file) and raise that instead of a `ValueError`.

File: pyahp/methods/power.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        iteration {number} -- [iteration] (default: {10})
+    
+    Returns:
+        eigenvalue, eigenvector
+    '''
+
+    m, n = A.shape
+    u0 = initvec or np.ones(n)
+    k = 0
+    while True:
```
Comment (mishrabhinav) [inline pyahp/methods/power.py]: Can we use the iteration count here to control the loop instead of `while True`?

File: pyahp/methods/power.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        v0 = u0
+        v1 = np.dot(A, v0)
+        ind = np.argmax(abs(v1))
+        mu = v1[ind]
+        u0 = v1 / mu
+        k += 1
+        if k >= iteration or LA.norm(v0 - u0)/np.max((1, LA.norm(u0))) < tol:
```
Comment (mishrabhinav) [inline pyahp/methods/power.py]: In relation to the previous comment, this could be a `for k in range(iteration)` instead of `while True`.

File: pyahp/utils.py [hunk 5]
```diff
@@ -21,3 +21,55 @@ def normalize_priorities(criteria_pr, global_pr):
...
+    Arguments:
+        A {2Darray|list(list)} -- The preference matrix
+    
+    Returns:
+        2Darray|list(list) -- Reciprocal matrix
+    '''
+
+    A = A.astype(np.float64)
+    m, n = A.shape
+    for i in range(m):
```
Comment (mishrabhinav) [inline pyahp/utils.py:42]: I am not going to lie but this loop looks terrible, so many levels of indentation! Can you try and reduce that?

File: pyahp/utils.py [hunk 6]
```diff
@@ -21,3 +21,55 @@ def normalize_priorities(criteria_pr, global_pr):
...
+                    for k in range(m):
+                        if A[i, k] !=0 and A[k, j] !=0:
+                            A[i, j] = A[i, k] * A[k, j]
+                            break
+                    else:
+                        raise ValueError('There are so many zeros! I am not able to convert A[%d,%d] to a vaild value!'%(i,j))
+    return A
+
+
+def is_reciprocal(A):
```
Comment (mishrabhinav) [inline pyahp/utils.py]: The `Method` abstract class has a function to check for reciprocal matrices already. Also, it doesn't make sense to add functions to the `utils` if they are not being used somewhere in the code.

File: examples/yettelevision.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "Television Model",
```
Comment (mishrabhinav) [inline examples/yettelevision.json]: This model could use a different and more descriptive name. Same goes for the filename.

File: pyahp/methods/power.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+This module contains the class implementing the geometric priority estimation method.
+"""
+
+import numpy as np
+import numpy.linalg as LA
+
+from pyahp.methods import Method
+
+
+def eig(A, initvec=None, tol=0.0001, iteration=10):
```
Comment (mishrabhinav) [inline pyahp/methods/power.py]: The implementation of the method is quite naive in my opinion as it seems to assume that it is always going to converge. I would like to see some test cases here with a good number of base, regular and edge cases.

**Review by mishrabhinav — CHANGES_REQUESTED** [review]
I like the idea of having zeroes instead of decimals in the preference matrices and it's something I would like to ship in the next release. For the power method, do you have any supporting text/paper that it is a reliable AHP priority calculation method?

File: examples/yettelevision.json (hunk 1 above)
Comment (Freakwill) [reply examples/yettelevision.json]: The Power Method is a classical method to calculate the single eigenvalue with maximal abstract value, i.e. |lambda_1| > |lambda2| >= |lambda_i| in numerical analysis. You can find it in any textbook about numerical analysis. The speed of convergence is dominated by |lambda2|/|lambda1|. Moreover, it is simpler then LM method.
Perron theorem states that lambda_1>|lambda_i| for any positive matrix.

Thank you for your advice.

File: examples/yettelevision.json (hunk 7 above)
Comment (Freakwill) [reply examples/yettelevision.json]: ok

File: pyahp/methods/power.py (hunk 3 above)
Comment (Freakwill) [reply pyahp/methods/power.py]: ok

File: pyahp/utils.py (hunk 2 above)
Comment (Freakwill) [reply pyahp/utils.py]: ok

Comment (Freakwill) [reply pyahp/utils.py]: I think it will be better that you define the error. Meanwhile I will think about it.

**Comment by Freakwill** [comment]
In the last commit, I considered your review as possible as I can. I don't kown modify it any more without your new review.

Do we need to define an interface for users? I am considering a fuzzy-version of AHP now. it seams uneasy.

I am attempting to use git commands at first.

File: pyahp/utils.py (hunk 5 above)
Comment (Freakwill) [reply pyahp/utils.py:42]: ok

File: pyahp/utils.py (hunk 6 above)
Comment (Freakwill) [reply pyahp/utils.py]: yes.

File: examples/yettelevision.json (hunk 1 above)
Comment (mishrabhinav) [reply examples/yettelevision.json]: Yes, I am very well aware of the power method to calculate the eigenvectors but my question is about the application of Power method to calculate the priorities from a preference matrix? I am looking for some kind of academic evidence here.

**Comment by mishrabhinav** [comment]
@Freakwill GitHub has great features and while it does allow you to upload files using the UI with a standard commit message, "Add files via upload", I would like you to change it to provide a more descriptive commit message. There are great guides online about writing commit messages and if you are interested, you could also look into using git from the command line([here](https://www.atlassian.com/git/tutorials)). Let me know if you need some help with that.

File: examples/yettelevision.json (hunk 1 above)
Comment (Freakwill) [reply examples/yettelevision.json]: `Perron theorem` is the evidence. The initial vector is recommended to be selected randomly.
